File: verifiers/typescript/src/tiers/tier4HonestCi.ts
// tier4HonestCi.ts — honest-CI reproducible-build driver.
//
// The full Tier 4 spec (docs/01-architecture/verifier-pipeline.md) wants:
//   1. Hermetic rebuild (Nix or Bazel), bit-identical hash compare
//   2. In-toto attestation signed by Sigstore keyless OIDC
//   3. SLSA L3 admissions
//   4. Rego policy admission
//
// Items 2/3/4 are owned by the verifier daemon (signing happens out of
// the per-language runner sandbox, by design — the runner has no OIDC
// identity). What we own here is item (1): run the project's build
// twice with `npm_config_prefer_frozen_lockfile=true`, sha256 the
// `dist/` tree of each run, compare.
//
// We deliberately use the TS-ecosystem-native `pnpm build` rather than
// shelling to Nix — Tier 4 ships with both paths and the TS runner
// implements the npm-side hermeticity check. The Nix rebuild is the
// daemon's job (apps/verifier/internal/honestci).

import { spawn } from "node:child_process";
import { createHash } from "node:crypto";
import { mkdtemp, readdir, readFile, rm, stat } from "node:fs/promises";
import os from "node:os";
import path from "node:path";

import {
  finalizeTestReport,
  newTestReport,
  type Finding,
  type HonestCIStats,
  type TestReport,
  type VerificationRequest,
} from "../schema.js";

export interface Tier4Options {
  cwd?: string;
  pnpmBin?: string;
  /** dir to sha256 after the build. Default: "dist". */
  outputDir?: string;
  wallClockSeconds?: number;
  /**
   * Inject the two hashes (test-only). When set, both build invocations
   * are skipped.
   */
  preSuppliedHashes?: { first: string; second: string };
  dryRun?: boolean;
}

// ---- Recursive tree hash ------------------------------------------------
//
// We hash the sorted list of (relative-path, sha256(content)) pairs. This
// gives us a deterministic hash independent of inode order or atime —
// the only sources of non-determinism we control. For npm packages, the
// remaining non-determinism (timestamps in package.json from `pnpm
// install`, etc.) is filtered: we only hash the `dist/` tree, which
// contains compiled outputs.

async function listFilesRecursive(root: string): Promise<string[]> {
  const out: string[] = [];
  async function walk(dir: string): Promise<void> {
    let entries;
    try {
      entries = await readdir(dir, { withFileTypes: true });
    } catch {
      return;
    }
    for (const e of entries) {
      const full = path.join(dir, e.name);
      if (e.isDirectory()) {
        await walk(full);
      } else if (e.isFile()) {
        out.push(full);
      }
    }
  }
  await walk(root);
  return out;
}

export async function hashDirectory(root: string): Promise<string> {
  const absRoot = path.resolve(root);
  const files = await listFilesRecursive(absRoot);
  files.sort();

  const tree = createHash("sha256");
  for (const f of files) {
    const rel = path.relative(absRoot, f).split(path.sep).join("/");
    let buf: Buffer;
    try {
      buf = await readFile(f);
    } catch {
      continue;
    }
    const fileHash = createHash("sha256").update(buf).digest("hex");
    tree.update(`${rel}\0${fileHash}\n`);
  }
  return tree.digest("hex");
}

// ---- Subprocess driver --------------------------------------------------

function runPnpm(
  bin: string,
  args: readonly string[],
  cwd: string,
  wallClockSeconds: number,
): Promise<{ exitCode: number; stderr: string }> {
  return new Promise((resolve, reject) => {
    const child = spawn(bin, Array.from(args), {
      cwd,
      stdio: ["ignore", "pipe", "pipe"],
      env: {
        ...process.env,
        npm_config_prefer_frozen_lockfile: "true",
        // Force deterministic behaviour where pnpm supports it.
        npm_config_progress: "false",
        FORCE_COLOR: "0",
        CI: "true",
      },
    });
    let stderr = "";
    let killed = false;
    const timer = setTimeout(
      () => {
        killed = true;
        child.kill("SIGKILL");
      },
      Math.max(1, wallClockSeconds) * 1000,
    );
    child.stderr?.on("data", (b: Buffer) => {
      stderr += b.toString("utf8");
      process.stderr.write(b);
    });
    child.stdout?.on("data", (b: Buffer) => {
      process.stderr.write(b);
    });
    child.on("error", (err) => {
      clearTimeout(timer);
      reject(err);
    });
    child.on("exit", (code) => {
      clearTimeout(timer);
      if (killed) {
        resolve({
          exitCode: 124,
          stderr: stderr + "\n[crucible: killed by wall-clock budget]",
        });
        return;
      }
      resolve({ exitCode: code ?? 1, stderr });
    });
  });
}

// ---- Public entrypoint --------------------------------------------------

export async function runTier4HonestCi(
  req: VerificationRequest,
  opts: Tier4Options = {},
): Promise<TestReport> {
  const startedAt = new Date();
  const cwd = opts.cwd ?? process.cwd();
  const outputDir = opts.outputDir ?? "dist";
  const wallClockSeconds =
    opts.wallClockSeconds ??
    Math.max(300, Math.min(1800, req.budget.wall_clock_cap_seconds || 300));

  const baseReport = newTestReport({
    task_id: req.task_id,
    diff_hash: req.diff.base_sha ?? req.base_sha,
    tier: "tier_4_honest_ci",
    framework: "pnpm-double-build",
    reporter_id: "crucible-verify-typescript",
    reporter_version: "2026.06.0-phase4",
    wall_clock_budget_seconds: wallClockSeconds,
    started_at: startedAt,
  });

  if (opts.preSuppliedHashes !== undefined) {
    const { first, second } = opts.preSuppliedHashes;
    const bitIdentical = first === second;
    const stats: HonestCIStats = {
      builder_id: "crucible://builders/pnpm-double-build/v1",
      executor_rebuild_hash: first,
      verifier_rebuild_hash: second,
      bit_identical: bitIdentical,
      slsa_level: bitIdentical ? 3 : 0,
      scrubber_audit_ok: true,
    };
    const findings: Finding[] = bitIdentical
      ? []
      : [
          {
            category: "rebuild_diverged",
            severity: "error" as const,
            detail: `dist/ hash diverged: first=${first}, second=${second}`,
          },
        ];
    return finalizeTestReport(
      {
        ...baseReport,
        honest_ci: stats,
        findings,
        verdict: bitIdentical ? "passed" : "failed",
        passed: bitIdentical,
      },
      new Date(),
    );
  }

  if (opts.dryRun === true) {
    return finalizeTestReport(
      {
        ...baseReport,
        verdict: "tool_unavailable",
        passed: false,
        error: "pnpm not invoked (dry-run)",
      },
      new Date(),
    );
  }

  const pnpm = opts.pnpmBin ?? "pnpm";

  // First pass — install + build into the project's own dist/.
  let firstStaging: string | null = null;
  let secondStaging: string | null = null;
  try {
    // We isolate each build to a temporary `--store-dir` to maximize
    // determinism; cwd remains the project root because the build
    // scripts likely use relative paths.
    firstStaging = await mkdtemp(path.join(os.tmpdir(), "crucible-tier4-pnpm1-"));
    secondStaging = await mkdtemp(path.join(os.tmpdir(), "crucible-tier4-pnpm2-"));

    const halfBudget = Math.max(1, Math.floor(wallClockSeconds / 2));

    const install1 = await runPnpm(
      pnpm,
      ["install", "--frozen-lockfile", `--store-dir=${firstStaging}`],
      cwd,
      halfBudget,
    );
    if (install1.exitCode !== 0) {
      return finalizeTestReport(
        {
          ...baseReport,
          verdict: "failed",
          passed: false,
          error: `pnpm install (pass 1) exit ${install1.exitCode}: ${install1.stderr.slice(-1000)}`,
        },
        new Date(),
      );
    }
    const build1 = await runPnpm(pnpm, ["build"], cwd, halfBudget);
    if (build1.exitCode !== 0) {
      return finalizeTestReport(
        {
          ...baseReport,
          verdict: "failed",
          passed: false,
          error: `pnpm build (pass 1) exit ${build1.exitCode}: ${build1.stderr.slice(-1000)}`,
        },
        new Date(),
      );
    }
    const distAbs = path.join(cwd, outputDir);
    try {
      await stat(distAbs);
    } catch {
      return finalizeTestReport(
        {
          ...baseReport,
          verdict: "failed",
          passed: false,
          error: `pnpm build (pass 1) produced no ${outputDir}/ directory`,
        },
        new Date(),
      );
    }
    const firstHash = await hashDirectory(distAbs);

    // Second pass.
    const install2 = await runPnpm(
      pnpm,
      ["install", "--frozen-lockfile", `--store-dir=${secondStaging}`],
      cwd,
      halfBudget,
    );
    if (install2.exitCode !== 0) {
      return finalizeTestReport(
        {
          ...baseReport,
          verdict: "failed",
          passed: false,
          error: `pnpm install (pass 2) exit ${install2.exitCode}: ${install2.stderr.slice(-1000)}`,
        },
        new Date(),
      );
    }
    const build2 = await runPnpm(pnpm, ["build"], cwd, halfBudget);
    if (build2.exitCode !== 0) {
      return finalizeTestReport(
        {
          ...baseReport,
          verdict: "failed",
          passed: false,
          error: `pnpm build (pass 2) exit ${build2.exitCode}: ${build2.stderr.slice(-1000)}`,
        },
        new Date(),
      );
    }
    const secondHash = await hashDirectory(distAbs);

    const bitIdentical = firstHash === secondHash;
    const stats: HonestCIStats = {
      builder_id: "crucible://builders/pnpm-double-build/v1",
      executor_rebuild_hash: firstHash,
      verifier_rebuild_hash: secondHash,
      bit_identical: bitIdentical,
      slsa_level: bitIdentical ? 3 : 0,
      scrubber_audit_ok: true,
    };
    const findings: Finding[] = bitIdentical
      ? []
      : [
          {
            category: "rebuild_diverged",
            severity: "error" as const,
            detail: `dist/ hash diverged: first=${firstHash}, second=${secondHash}`,
          },
        ];

    return finalizeTestReport(
      {
        ...baseReport,
        honest_ci: stats,
        findings,
        verdict: bitIdentical ? "passed" : "failed",
        passed: bitIdentical,
      },
      new Date(),
    );
  } catch (err) {
    return finalizeTestReport(
      {
        ...baseReport,
        verdict: "tool_unavailable",
        passed: false,
        error: `tier 4 driver failed: ${(err as Error).message}`,
      },
      new Date(),
    );
  } finally {
    for (const dir of [firstStaging, secondStaging]) {
      if (dir !== null) {
        void rm(dir, { recursive: true, force: true }).catch(() => {});
      }
    }
  }
}

export const __testing__ = { hashDirectory };

File: verifiers/rust/src/tiers/tier4_honest_ci.rs
//! Tier-4 honest-CI rebuild check.
//!
//! We run `cargo build --release` twice with `SOURCE_DATE_EPOCH=0` and,
//! when the compiler supports it, `-Z trim-paths` (stable equivalent on
//! recent toolchains is `RUSTFLAGS=-Cdebuginfo=0 --remap-path-prefix`).
//! After each pass we SHA-256 the produced executable; `bit_identical`
//! is set when the two digests match.
//!
//! This is a per-language sub-check; the full Tier-4 SLSA-L3 attestation
//! is composed in the Go dispatcher (`apps/verifier/internal/tier4/`).
//! We populate the `HonestCIStats` fields that are Rust-rebuild-specific
//! and leave the dispatcher to fill in the Rekor/Fulcio/Witness fields.

use std::path::{Path, PathBuf};
use std::process::Command;
use std::time::Instant;

use sha2::{Digest, Sha256};

use crate::schema::{HonestCiStats, TestReport, Verdict, VerificationRequest};
use crate::tiers::locate;

/// Marker baked into the report so dispatcher-side joiners know which
/// rebuilder produced the digest pair.
pub const BUILDER_ID: &str = "https://crucible.dev/builders/rust-cargo-double-build/v1";

/// Entrypoint invoked by `tiers::dispatch`.
pub fn run(_req: &VerificationRequest, mut report: TestReport) -> TestReport {
    report.framework = "cargo-double-build".to_string();
    let started = Instant::now();

    let Some(cargo) = locate("cargo") else {
        report.stamp_finished();
        return report.tool_unavailable("cargo not on PATH");
    };

    let work = match tempfile::tempdir() {
        Ok(t) => t,
        Err(e) => {
            report.stamp_finished();
            return report.tool_unavailable(format!("tempdir: {e}"));
        }
    };

    let a_target = work.path().join("target-a");
    let b_target = work.path().join("target-b");

    let a_hash = match build_once(&cargo, &a_target) {
        Ok(h) => h,
        Err(e) => {
            report.stamp_finished();
            return report.tool_unavailable(format!("first cargo build: {e}"));
        }
    };
    let b_hash = match build_once(&cargo, &b_target) {
        Ok(h) => h,
        Err(e) => {
            report.stamp_finished();
            return report.tool_unavailable(format!("second cargo build: {e}"));
        }
    };

    let bit_identical = a_hash == b_hash;
    let stats = HonestCiStats {
        builder_id: BUILDER_ID.to_string(),
        executor_rebuild_hash: a_hash,
        verifier_rebuild_hash: b_hash,
        bit_identical,
        // SLSA level for a Rust-only double-build (no in-toto, no
        // Sigstore signing) is L2 at best — Go dispatcher upgrades to L3
        // when the rebuild lands inside its hardened Witness pipeline.
        slsa_level: 2,
        scrubber_audit_ok: true,
        ..HonestCiStats::default()
    };

    report.verdict = if bit_identical {
        Verdict::Passed
    } else {
        Verdict::Failed
    };
    report.passed = bit_identical;
    report.honest_ci = Some(stats);
    report.duration_seconds = started.elapsed().as_secs_f64();
    report.stamp_finished();
    report
}

/// Run a single `cargo build --release` into the supplied target dir.
/// Returns the hex-encoded SHA-256 of the produced primary binary.
fn build_once(cargo: &Path, target_dir: &Path) -> anyhow::Result<String> {
    std::fs::create_dir_all(target_dir)?;
    // Pass --release into a dedicated target dir; identical
    // SOURCE_DATE_EPOCH and remap-path-prefix scrubs the timestamp +
    // build-host path stamps Cargo otherwise embeds.
    let cwd = std::env::current_dir()?;
    let cwd_str = cwd.to_string_lossy();
    let cwd_trimmed = cwd_str.trim_end_matches('/').trim_end_matches('\\');
    let remap = format!("--remap-path-prefix={cwd_trimmed}=.");
    let rustflags = format!("-C debuginfo=0 {remap}");
    let status = Command::new(cargo)
        .arg("build")
        .arg("--release")
        .arg("--target-dir")
        .arg(target_dir)
        .env("SOURCE_DATE_EPOCH", "0")
        .env("RUSTFLAGS", &rustflags)
        .env("CARGO_TERM_COLOR", "never")
        .status()?;
    anyhow::ensure!(status.success(), "cargo build exit {:?}", status.code());

    let bin_dir = target_dir.join("release");
    let primary = pick_primary_artifact(&bin_dir)?;
    hash_file(&primary)
}

/// Pick the most-recently-modified executable / library artifact under
/// `target/release/`. We avoid `target/release/deps/` and Cargo's
/// `.d`/`.json` sidecars.
fn pick_primary_artifact(release_dir: &Path) -> anyhow::Result<PathBuf> {
    let rd = std::fs::read_dir(release_dir)
        .map_err(|e| anyhow::anyhow!("read release dir {release_dir:?}: {e}"))?;
    let mut best: Option<(std::time::SystemTime, PathBuf)> = None;
    for entry in rd.filter_map(Result::ok) {
        let path = entry.path();
        if !path.is_file() {
            continue;
        }
        // Skip Cargo sidecar files.
        let name = path.file_name().and_then(|s| s.to_str()).unwrap_or("");
        if name.ends_with(".d")
            || name.ends_with(".json")
            || name.starts_with('.')
            || name.ends_with(".rlib")
        {
            continue;
        }
        let meta = entry.metadata().ok();
        let mtime = meta
            .and_then(|m| m.modified().ok())
            .unwrap_or(std::time::UNIX_EPOCH);
        if best.as_ref().map_or(true, |b| mtime > b.0) {
            best = Some((mtime, path));
        }
    }
    best.map(|(_, p)| p)
        .ok_or_else(|| anyhow::anyhow!("no artifact in {release_dir:?}"))
}

fn hash_file(path: &Path) -> anyhow::Result<String> {
    let bytes = std::fs::read(path)?;
    let mut h = Sha256::new();
    h.update(&bytes);
    Ok(hex::encode(h.finalize()))
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::io::Write;

    #[test]
    fn hash_file_is_stable() {
        let mut f = tempfile::NamedTempFile::new().unwrap();
        f.write_all(b"crucible").unwrap();
        let a = hash_file(f.path()).unwrap();
        let b = hash_file(f.path()).unwrap();
        assert_eq!(a, b);
        assert_eq!(a.len(), 64);
    }

    #[test]
    fn picks_primary_artifact_over_sidecar() {
        let dir = tempfile::tempdir().unwrap();
        std::fs::write(dir.path().join("foo.d"), b"deps").unwrap();
        std::fs::write(dir.path().join("foo"), b"binary").unwrap();
        let picked = pick_primary_artifact(dir.path()).unwrap();
        assert!(picked.ends_with("foo"));
    }
}
